src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+        Exchange::new(move |_| hashed_id),
+        vec![],
+    );
+    let mut batches_input =
+        append_op.new_input_connection(batches, Exchange::new(move |_| hashed_id), vec![]);
+
+    let current_upper = Rc::clone(&storage_state.source_uppers[&collection_id]);
+    if !active_worker {
+        // This worker is not writing, so make sure it's "taken out" of the
+        // calculation by advancing to the empty frontier.
+        current_upper.borrow_mut().clear();
+    }
+
+    let source_statistics = storage_state
+        .source_statistics
+        .get(&collection_id)
+        .expect("statistics initialized")
+        .clone();
+
+    // This operator accepts the batch descriptions and tokens that represent
+    // written batches. Written batches get appended to persist when we learn
+    // from our input frontiers that we have seen all batches for a given batch
+    // description.
+
+    let shutdown_button = append_op.build(move |_| async move {
+        // This may SEEM unnecessary, but metrics contains extra
+        // `DeleteOnDrop`-wrapped fields that will NOT be moved into this
+        // closure otherwise, dropping and destroying
+        // those metrics. This is because rust now only moves the
+        // explicitly-referenced fields into closures.
+        let metrics = metrics;
+
+        // Contains descriptions of batches for which we know that we can
+        // write data. We got these from the "centralized" operator that
+        // determines batch descriptions for all writers.
+        //
+        // `Antichain` does not implement `Ord`, so we cannot use a `BTreeSet`. We need to search
+        // through the set, so we cannot use the `mz_ore` wrapper either.
+        #[allow(clippy::disallowed_types)]
+        let mut in_flight_descriptions =
+            std::collections::HashSet::<(Antichain<Timestamp>, Antichain<Timestamp>)>::new();
+
+        // In flight batches that haven't been `compare_and_append`'d yet, plus metrics about
+        // the batch.
+        let mut in_flight_batches = HashMap::<
+            (Antichain<Timestamp>, Antichain<Timestamp>),
+            Vec<(Batch<_, _, _, _>, BatchMetrics)>,
+        >::new();
+
+        if !active_worker {
+            // The non-active workers report that they are done snapshotting.
+            source_statistics.initialize_snapshot_committed(&Antichain::<Timestamp>::new());
+            return;
+        }
+
+        // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+        let persist_client = persist_clients
+            .open(persist_location)
+            .await
+            .expect("could not open persist client");
+
+        let mut write = persist_client
+            .open_writer::<SourceData, (), Timestamp, Diff>(
+                shard_id,
+                &format!("persist_sink::append_batches {}", collection_id),
+                Arc::new(target_relation_desc),
+                Arc::new(UnitSchema),
```

Comment:
We should leave a `TODO` here to add the correct schema

Reply:
This is the value schema, which is always empty